index.js

```diff
@@ -12,7 +12,26 @@ function isNegative(pattern) {
 	return pattern[0] === '!';
 }
 
+function isString(value) {
+	return typeof value === 'string';
+}
+
+function assertPatternsInput(patterns) {
+	if (isString(patterns)) {
+		// bail early, the rest will easier this way
+		return;
+	}
+
+	if (!Array.isArray(patterns) || !patterns.every(isString)) {
+		throw new TypeError('patterns must be a string or an array of strings');
+	}
+}
+
 function generateGlobTasks(patterns, opts) {
+	// make sure input is correct, so that we can safely parse
+	// it in this task
+	assertPatternsInput(patterns);
+
 	var globTasks = [];
 
 	patterns = arrify(patterns);
```

Comment:
I am not actually sure what is being suggested here. Are you saying that `arrify` should be used before validation, and the values `null` and `undefined` should be allowed through?

The problem I have there is that `null` and `undefined` would result in a successful operation returning `[]`, but `[null]` and `[undefined]` will result in a `TypeError`. The type for patters is a string or array of strings. `null` and `undefined` are valid types with are not a string or array of strings. Yes, I agree that it is less code, but it is also more unexpected behavior.